src/SelfGravity/CPU_Poisson/CPU_ExtPot_GREP.cpp

```diff
@@ -0,0 +1,256 @@
+   AuxArray[5] = GREPSgTime[ FaLv ][     Sg_FaLv ]; // new physical time of GREP on father  level
+   AuxArray[6] = GREPSgTime[ FaLv ][ 1 - Sg_FaLv ]; // old physical time of GREP on father  level
+
+} // FUNCTION : SetExtPotAuxArray_GREP
+#endif // #ifndef __CUDACC__
+
+
+
+// =================================
+// II. Specify external potential
+// =================================
+
+//-----------------------------------------------------------------------------------------
+// Function    :  ExtPot_GREP
+// Description :  Calculate the external potential at the given coordinates and time
+//
+// Note        :  1. This function is shared by CPU and GPU
+//                2. Auxiliary array UserArray[] is not used here
+//                3. Currently it does not support the soften length
+//
+// Parameter   :  x/y/z     : Target spatial coordinates
+//                Time      : Target physical time
+//                UserArray : User-provided auxiliary array
+//                Usage     : Different usages of external potential when computing total potential on level Lv
+//                            --> EXT_POT_USAGE_ADD     : add external potential on Lv
+//                                EXT_POT_USAGE_SUB     : subtract external potential for preparing self-gravity potential on Lv-1
+//                                EXT_POT_USAGE_SUB_TINT: like SUB but for temporal interpolation
+//                            --> This parameter is useless in most cases
+//
+// Return      :  External potential at (x,y,z,Time)
+//-----------------------------------------------------------------------------------------
+GPU_DEVICE_NOINLINE
+static real ExtPot_GREP( const double x, const double y, const double z, const double Time, const double UserArray[],
+                         const ExtPotUsage_t Usage )
+{
+
+   int     NBin;
+   double  pot;
+   double *effpot;
+   double *radius;
+
+   const real dx = (real)( x - UserArray[0] );
+   const real dy = (real)( y - UserArray[1] );
+   const real dz = (real)( z - UserArray[2] );
+   const real r  = SQRT( SQR(dx) + SQR(dy) + SQR(dz) );
+
+
+// use Usage to determine which Data and Radius profiles, and NBin are used
+#ifdef __CUDACC__
+   if ( Usage == EXT_POT_USAGE_ADD )
+   {
+      effpot = c_GREP_Lv_Data_New;
+      radius = c_GREP_Lv_Radius_New;
+      NBin   = c_GREP_Lv_NBin_New;
+   }
+#else // #ifdef __CUDACC__
+   switch ( Usage )
+   {
+      case EXT_POT_USAGE_ADD:
+         effpot = h_GREP_Lv_Data_New;
+         radius = h_GREP_Lv_Radius_New;
+         NBin   = h_GREP_Lv_NBin_New;
+      break;
+
+      case EXT_POT_USAGE_SUB:
+         effpot = h_GREP_FaLv_Data_New;
+         radius = h_GREP_FaLv_Radius_New;
+         NBin   = h_GREP_FaLv_NBin_New;
+      break;
+
+      case EXT_POT_USAGE_SUB_TINT:
+         effpot = h_GREP_FaLv_Data_Old;
+         radius = h_GREP_FaLv_Radius_Old;
+         NBin   = h_GREP_FaLv_NBin_Old;
+      break;
```

Comment:
By checking `Poi_Prepare_Pot()` and `Poi_Prepare_GREP()`, I suppose it should be correct. However, it could not the case at the beginning of simulation. Now I compare the input time with the time in `AuxArray` to choose the correct one.